### PR TITLE
Adding cadcading deletion support for federated secrets

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/33612

Adding cascading deletion support for federated secrets.
The code is same as that for namespaces.  Just ensuring that DeletionHelper functions are called at right places in secret_controller.
Also added e2e tests.

cc @kubernetes/sig-cluster-federation @caesarxuchao

```release-note
federation: Adding support for DeleteOptions.OrphanDependents for federated secrets. Setting it to false while deleting a federated secret also deletes the corresponding secrets from all registered clusters.
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36296)

### CODE REVIEW

File: federation/pkg/federation-controller/secret/secret_controller.go [hunk 1]
```diff
@@ -316,3 +410,24 @@ func (secretcontroller *SecretController) reconcileSecret(secret types.Namespace
 	// Evertyhing is in order but lets be double sure
 	secretcontroller.deliverSecret(secret, secretcontroller.secretReviewDelay, false)
 }
+
+// delete deletes the given secret or returns error if the deletion was not complete.
+func (secretcontroller *SecretController) delete(secret *api_v1.Secret) error {
+	glog.V(3).Infof("Handling deletion of secret: %v", *secret)
+	// Delete the secret from all underlying clusters.
```
Comment (caesarxuchao) [inline federation/pkg/federation-controller/secret/secret_controller.go]: Update the comment.

File: federation/pkg/federation-controller/secret/secret_controller_test.go [hunk 2]
```diff
@@ -92,6 +95,11 @@ func TestSecretController(t *testing.T) {
 
 	// Test add federated secret.
 	secretWatch.Add(&secret1)
+	updatedSecret := GetSecretFromChan(secretUpdateChan)
+	assert.True(t, secretController.hasFinalizerFunc(updatedSecret, deletionhelper.FinalizerDeleteFromUnderlyingClusters))
```
Comment (caesarxuchao) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Check for both finalizers.

**Comment by caesarxuchao** [comment]
LGTM. Update the comments, and update the assert in the e2e test. Feel free to self apply the label.

File: federation/pkg/federation-controller/secret/secret_controller_test.go (hunk 2 above)
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller_test.go]: Done

File: federation/pkg/federation-controller/secret/secret_controller.go (hunk 1 above)
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller.go]: Removed it

File: federation/pkg/federation-controller/secret/secret_controller.go [hunk 3]
```diff
@@ -248,6 +316,32 @@ func (secretcontroller *SecretController) reconcileSecret(secret types.Namespace
...
+			secretcontroller.deliverSecret(secret, 0, true)
+		}
+		return
+	}
+
+	glog.V(3).Infof("Ensuring delete object from underlying clusters finalizer for secret: %s",
+		baseSecret.Name)
+	// Add the required finalizers before creating a secret in
+	// underlying clusters.
+	// This ensures that the dependent secrets are deleted in underlying
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/secret/secret_controller.go]: Done

**Comment by nikhiljindal** [comment]
Updated the code as per comments.
Thanks a lot @caesarxuchao for the review!

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
Rebased

**Comment by saad-ali** [comment]
Release-czar approved post-code freeze merge--This was LGTMed and in the merge-queue at code freeze time for 1.5. Leaving 1.5 milestone to let it gets merged after code freeze.

**Comment by rmmh** [comment]
@nikhiljindal @pwittrock this is making test-go fail ~25% of the time (#36422). Can this be reverted if it's not an easy fix?

**Comment by nikhiljindal** [comment]
Sorry just saw https://github.com/kubernetes/kubernetes/issues/36422. Am on it

**Pull request merged**